src/audio/google/google_rtc_audio_processing.c

```diff
@@ -83,31 +112,236 @@ void GoogleRtcFree(void *ptr)
-	component_set_nearest_period_frames(dev, params->rate);
+	if (!bmec)
+		return -EINVAL;
+
+	bce = &bmec->base_cfg_ext;
+	if (bce->nb_input_pins != 2 && bce->nb_output_pins != 1) {
+		comp_err(mod->dev, "Must have 2 input, 1 output pins");
+		return -EINVAL;
+	}
+
+	ipf = (void *)&bce->pin_formats[0];
+	for (i = 0; i < bce->nb_input_pins; i++, ipf++) {
+		if (ipf->pin_index < 0 || ipf->pin_index >= 2) {
+			comp_err(mod->dev, "Incorrect input pin index %d", ipf->pin_index);
+			return -EINVAL;
+		}
+		cd->in_fmts[ipf->pin_index] = ipf->audio_fmt;
+	}
 
-	list_for_item(source_list, &dev->bsource_list) {
-		sourceb = container_of(source_list, struct comp_buffer, sink_list);
-		if (IPC4_SINK_QUEUE_ID(buf_get_id(sourceb)) == SOF_AEC_FEEDBACK_QUEUE_ID)
-			ipc4_update_buffer_format(sourceb, &cd->config.reference_fmt);
-		else
-			ipc4_update_buffer_format(sourceb, &mod->priv.cfg.base_cfg.audio_fmt);
+	opf = (void *)ipf;
+	if (opf->pin_index != 0) {
+		comp_err(mod->dev, "Incorrect output pin index %d\n", opf->pin_index);
+		return -EINVAL;
 	}
+	cd->out_fmt = opf->audio_fmt;
+
+	size_t bcesz = sizeof(*bce) + (bce->nb_input_pins * sizeof(*ipf)
+				       + bce->nb_output_pins * sizeof(*opf));
+	void *bcedup = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, bcesz);
+
+	if (!bcedup)
+		return -ENOMEM;
+	memcpy(bcedup, bce, bcesz);
+	mod->priv.cfg.basecfg_ext = bcedup;
 
-	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
-	ipc4_update_buffer_format(sinkb, &mod->priv.cfg.base_cfg.audio_fmt);
+	return 0;
+}
+
+static void prepare_ipc4_fmts(struct processing_module *mod,
+			      struct sof_source **sources,
+			      struct sof_sink **sinks)
+{
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	/* FIXME: The "pin" index (the value of the pin_index field of
+	 * the ipc4_in/output_pin_format structs seen in module config
+	 * at init() time) and the "source" index (the ordering of the
+	 * sources[] argument to prepare()/process()) ARE PRESENTED
+	 * BACKWARDS!  And I don't see any API to tell me which is
+	 * which (note that the ordering of the connected buffers in
+	 * the comp_dev sink/source lists is a THIRD convention, and
+	 * matches sources[]/sinks[]).
+	 */
+	ipc4_update_source_format(sources[0], &cd->in_fmts[1]);
+	ipc4_update_source_format(sources[1], &cd->in_fmts[0]);
+	ipc4_update_sink_format(sinks[0], &cd->out_fmt);
+
+	source_set_alignment_constants(sources[0], 1, 1);
+	source_set_alignment_constants(sources[1], 1, 1);
+	sink_set_alignment_constants(sinks[0], 1, 1);
 }
 #endif
 
+static ALWAYS_INLINE float s16_to_float(const char *ptr)
+{
+	float scale = -(float)SHRT_MIN;
+	float x = *(int16_t *)ptr;
+
+	return (1.0f / scale) * x;
+}
+
+static ALWAYS_INLINE void float_to_s16(float x, char *dst)
+{
+	float scale = -(float)SHRT_MIN;
+	float min = -1.0f;
+	float max = 1.0f - 1.0f / scale;
+	int16_t i = (int16_t)(scale * (x < min ? min : (x > max ? max : x)));
```

Comment:
I find more readable the comparison of `scale * x` with SHRT_MIN and SHRT_MAX. It is also more likely to be more robust as they won't be anyy rounding afterward.
```suggestion
    const float v = scale * x;
	int16_t i = v   < (float) SHRT_MIN?SHRT_MIN : v > (float) SHRT_MAX ? SHRT_MAX : v))
```